chain/vm/fvm.go

```diff
@@ -0,0 +1,312 @@
+package vm
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"github.com/filecoin-project/go-state-types/network"
+
+	"github.com/filecoin-project/go-state-types/big"
+
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/state"
+	cbor "github.com/ipfs/go-ipld-cbor"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/exitcode"
+	"github.com/filecoin-project/lotus/lib/sigs"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/blockstore"
+
+	ffi "github.com/filecoin-project/filecoin-ffi"
+	ffi_cgo "github.com/filecoin-project/filecoin-ffi/cgo"
+
+	"github.com/filecoin-project/lotus/chain/actors/adt"
+	"github.com/filecoin-project/lotus/chain/actors/builtin/miner"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/ipfs/go-cid"
+)
+
+var _ Interface = (*FVM)(nil)
+var _ ffi_cgo.Externs = (*FvmExtern)(nil)
+
+type FvmExtern struct {
+	Rand
+	blockstore.Blockstore
+	epoch   abi.ChainEpoch
+	lbState LookbackStateGetter
+	base    cid.Cid
+}
+
+// VerifyConsensusFault is similar to the one in syscalls.go used by the LegacyVM, except it never errors
+// Errors are logged and "no fault" is returned, which is functionally what go-actors does anyway
+func (x *FvmExtern) VerifyConsensusFault(ctx context.Context, a, b, extra []byte) (*ffi_cgo.ConsensusFault, int64) {
```

Comment:
What happens if any of those extern calls panic? Do we crash the node, or do we behave like the current VM and just fail execution?

Reply:
@magik6k Thanks a lot for raising this. This is a somewhat open question about exactly what we wanna do -- currently:

- Rust panics get converted to errors at the FFI layer
- Extern panics such as this one are...unclear? Maybe also get lowered to errors at the FFI laye, maybe crash the node? @Stebalien isn't sure, we can relatively easy answer this by experimenting.
- Any error returned by the FFI (including such downgraded panics) cause the `ApplyMessage` call to fail and the node to freeze. In terms of the `LegacyVM`, this means all such errors are "fatal errors". I'm fairly opposed to this, but a decision will be made later.

Reply:
> Extern panics such as this one are...unclear? Maybe also get lowered to errors at the FFI laye, maybe crash the node? @Stebalien isn't sure, we can relatively easy answer this by experimenting.

The Rust => Go call path is brand new, so this definitely deseves a ton of attention.